openstack/pkg/designate/provider.go

```diff
@@ -0,0 +1,161 @@
+package designate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/dns/v2/recordsets"
+	"github.com/libdns/libdns"
+)
+
+// Provider implements the libdns interfaces for OpenStack Designate.
+type Provider struct {
+	dnsClient      *gophercloud.ServiceClient
+	AuthOpenStack  AuthOpenStack `yaml:"auth_open_stack"`
+	zoneID         string
+}
+
+// AuthOpenStack contains credentials for OpenStack Designate.
+type AuthOpenStack struct {
+	RegionName         string `yaml:"region_name"`
+	TenantID           string `yaml:"tenant_id"`
+	IdentityApiVersion string `yaml:"identity_api_version"`
+	Password           string `yaml:"password"`
+	AuthURL            string `yaml:"auth_url"`
+	Username           string `yaml:"username"`
+	TenantName         string `yaml:"tenant_name"`
+	EndpointType       string `yaml:"endpoint_type"`
```

Comment:
Add `json` struct tags too if you could. (The majority of users, Caddy users, will need that.)